disorder-jack/src/Disorder/Jack/Tree.hs

```diff
@@ -0,0 +1,160 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+module Disorder.Jack.Tree (
+    Tree(..)
+  , foldTree
+  , foldForest
+  , unfoldTree
+  , unfoldForest
+  , filterTree
+  , filterForest
+  , expandTree
+  ) where
+
+import           Control.Applicative (Applicative(..))
+import           Control.Comonad (Comonad(..), ComonadApply(..))
+import           Control.DeepSeq (NFData(..))
+import           Control.Monad (Monad(..))
+
+import           Data.Bool (Bool)
+import           Data.Data (Data)
+import           Data.Eq (Eq)
+import           Data.Foldable (Foldable(..))
+import           Data.Function (($), (.), id)
+import           Data.Functor (Functor(..), (<$>))
+import qualified Data.List as List
+import           Data.Monoid ((<>))
+import           Data.Ord (Ord)
+import           Data.Traversable (Traversable(..))
+import           Data.Typeable (Typeable)
+
+import           GHC.Generics (Generic)
+
+import           Prelude (seq)
+
+import           Text.Show (Show)
+
+
+-- | A rose tree which represents a random generated outcome, and all the ways
+--   in which it can made smaller.
+--
+--   This tree is exactly the same as 'Data.Tree' in every way except that
+--   Applicative '<*>' and Monad '>>=' walk the tree in the reverse order. This
+--   modification is critical for shrinking to reach a minimal counterexample.
+--
+data Tree a =
+  Node {
+    -- | The generated outcome.
+    outcome :: !a
+
+    -- | All the possible shrinks of this outcome.
+  , shrinks :: [Tree a]
```

Comment:
It's possibly more important to just not that we don't backtrack. So you want smallest first because it means we terminate the search quicker, but you don't want it at the expense of not trying paths which might end up being smaller in the end. This is why `expandTree` adds the extra nodes to the end, I couldn't figure out how to apply the shrink function "everywhere" so to speak.